### PR TITLE
Change host relay tests to check error returned

### PR DESCRIPTION
**Pull request by DimitrisJim**
## Description

closes: #4145 

### Commit Message / Changelog Entry

```bash
chore: change host relay tests to check error returned
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go [hunk 1]
```diff
@@ -330,6 +330,40 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
 			},
 			true,
 		},
+		{
+			"Msg fails its ValidateBasic: MsgTransfer has an empty receiver",
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/keeper/relay_test.go:334]: tested locally by removing `ValidateBasic`, this doesn't actually fix the issue, because I believe it will fail to execute. We need a message which fails validate basic but succeeds in execution

Comment (colin-axner) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:334]: I wonder if we should first open an issue to convert this test to assert the error returned

Comment (DimitrisJim) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:334]: gotcha, let me look into it.

Comment (DimitrisJim) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:334]: ya looked into this a bit, we get the same error even if we convert to assert the error. `executeMsg` just goes through the normal Msg processing flow and will eventually also call `ValidateBasic` on the faulty message we create:

https://github.com/cosmos/cosmos-sdk/blob/095a64195aba917187bbb33570063bc675494bcd/baseapp/msg_service_router.go#L130-L134

Comment (colin-axner) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:334]: But wouldn't the returned error change if the validate basic check was removed?

File: modules/apps/27-interchain-accounts/types/codec.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ func DeserializeCosmosTx(cdc codec.Codec, data []byte, encoding string) ([]sdk.M
 	switch encoding {
 	case EncodingProtobuf:
 		if err := cdc.Unmarshal(data, &cosmosTx); err != nil {
-			return nil, errorsmod.Wrapf(err, "cannot unmarshal CosmosTx with protobuf")
+			return nil, errorsmod.Wrapf(ErrUnknownDataType, "cannot unmarshal CosmosTx with protobuf")
```
Comment (DimitrisJim) [inline modules/apps/27-interchain-accounts/types/codec.go:87]: I can't see a reason why this was returning the raw unmarshal error? Lmk if I'm missing something here.

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go (hunk 1 above)
Comment (DimitrisJim) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:335]: resolved in slack discussion: see parent issue edit for resolution.

**Comment by DimitrisJim** [comment]
hm, I'm also going to change this for the JSON tests real quick.

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go [hunk 3]
```diff
@@ -494,11 +529,12 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
 
 				txResponse, err := suite.chainB.GetSimApp().ICAHostKeeper.OnRecvPacket(suite.chainB.GetContext(), packet)
 
-				if tc.expPass {
-					suite.Require().NoError(err)
+				expPass := tc.expErr == nil
+				if expPass {
+					suite.Require().Nil(err)
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/keeper/relay_test.go:534]: just curious, why the change from `NoError` to `Nil`?

File: modules/apps/27-interchain-accounts/types/codec.go (hunk 2 above)
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/types/codec.go:87]: `ErrUnknownDataType` is somewhat confusing to me here. It failed to unmarshal which could be from malformed data

> I can't see a reason why this was returning the raw unmarshal error?

Probably just based on the pattern of, "if you see an error, wrap it". I'm okay using our own error as a base, but we should be emitting the unmarshal information in the err msg

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go (hunk 3 above)
Comment (DimitrisJim) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:534]: I'm pretty sure `tc.expErr == nil` moved me in that direction as I was reading it. Looking at others I do see `NoError` preferred so I can swap to that.

File: modules/apps/27-interchain-accounts/types/codec.go (hunk 2 above)
Comment (DimitrisJim) [reply modules/apps/27-interchain-accounts/types/codec.go:87]: yea, that sounds sensible to me. 

was also slightly confused but went with what was currently used for json unmarhalling, this error seems to be slightly overused :smiley: could add a new one in a follow up

**Review by colin-axner — APPROVED** [review]
Thanks @DimitrisJim!!

**Review by chatton — APPROVED** [review]
LGTM nice cleanup 👍

**Review by damiannolan — APPROVED** [review]
Good stuff! Thanks @DimitrisJim 🚀

**Review by charleenfei — APPROVED** [review]
thanks for the cleanup!
